torch/testing/_internal/opinfo_helper.py

```diff
@@ -0,0 +1,110 @@
+from collections import defaultdict
+import warnings
+from functools import partial
+
+import torch
+from torch.testing._internal.common_cuda import (TEST_CUDA)
+from torch.testing._internal.common_methods_invocations import (op_db, _DYNAMIC_DTYPES)
+from torch.testing import (all_types_and_complex_and,
+                           all_types_and_complex,
+                           # all_types_and_half, # not exported by testing at the moment
+                           all_types,
+                           complex_types,
+                           floating_and_complex_types,
+                           # floating_types_and_half, # not exported by testing at the moment
+                           floating_types,
+                           integral_types,
+                           )
+
+DTYPE_DISPATCH = (
+    floating_types,
+    floating_and_complex_types,
+    integral_types,
+    all_types,
+    complex_types,
+    all_types_and_complex,
+)
+
+# Better way to acquire devices?
+DEVICES = ['cpu'] + (['cuda'] if TEST_CUDA else [])
```

Comment:
We should create a better way. cc @pmeier 

I think we should add function to torch.testing that returns the testing devices. This function can be implemented by acquiring

https://github.com/pytorch/pytorch/blob/fa0b89bbf7eb8e106930accc5324f47b2a76370f/torch/testing/_internal/common_device_type.py#L431

This internal data structure is populated by 

https://github.com/pytorch/pytorch/blob/fa0b89bbf7eb8e106930accc5324f47b2a76370f/torch/testing/_internal/common_device_type.py#L189

which has a function

https://github.com/pytorch/pytorch/blob/fa0b89bbf7eb8e106930accc5324f47b2a76370f/torch/testing/_internal/common_device_type.py#L210

which is typically the device type (e.g. for the CPU test base it returns 'cpu') but not for CUDA:

https://github.com/pytorch/pytorch/blob/fa0b89bbf7eb8e106930accc5324f47b2a76370f/torch/testing/_internal/common_device_type.py#L380

where it is the current CUDA device:

https://github.com/pytorch/pytorch/blob/fa0b89bbf7eb8e106930accc5324f47b2a76370f/torch/testing/_internal/common_device_type.py#L404

This is why, by the way, the test suite shouldn't have checks like `device == 'cuda'`, because the CUDA primary device is an actual device (specified with a string like `cuda:0` or `cuda:1`).

Anyway, this function in torch.testing can acquire the test bases and then query for their device types and devices to run these tests on.

Reply:
@mruberry recall that we had problems when importing anything from `torch.testing._internal`. This is why we duplicated https://github.com/pytorch/pytorch/blob/74c12da4517c789bea737dc947d6adc755f63176/torch/testing/_asserts.py#L34-L46

So we probably need to duplicate the functionality for now.

Reply:
Other than that, I agree: something like that would be a good addition.

Reply:
I've added this to our tracker